File: packages/i18nify-js/src/modules/names/isValidNameFormat.ts
import { withErrorBoundary } from '../../common/errorBoundary';

/**
 * Validates the format of a personal name against i18nify name rules.
 *
 * A name has a valid format when it:
 * - Is at least 2 characters long (after trimming whitespace).
 * - Is at most 100 characters long.
 * - Contains only Unicode letters, spaces, hyphens (-), apostrophes ('), or periods (.).
 * - Contains at least one Unicode letter.
 *
 * This is a structural/charset check. For garbage-input detection (blocklists,
 * sequential/repeating characters), use `isValidName` from the nameValidation module.
 *
 * @param {string} name - The name to validate
 * @returns {boolean} True if the name's format is valid, false otherwise
 *
 * @example
 * isValidNameFormat('John Smith');    // true
 * isValidNameFormat("O'Brien");       // true
 * isValidNameFormat('Mary-Jane');     // true
 * isValidNameFormat('John2');         // false — digits not allowed
 * isValidNameFormat('A');             // false — too short
 */
const isValidNameFormat = (name: string): boolean => {
  const trimmed = (name ?? '').trim();

  const MIN_LENGTH = 2;
  const MAX_LENGTH = 100;

  if (trimmed.length < MIN_LENGTH || trimmed.length > MAX_LENGTH) {
    return false;
  }

  // Must contain at least one Unicode letter and only allowed characters.
  // \p{M} includes combining marks (e.g., Devanagari vowel signs, Arabic diacritics)
  // which are integral parts of correctly-spelled names in many scripts.
  // Literal space is used (not \s) to exclude tabs and newlines.
  const allowedPattern = /^[\p{L}\p{M} '\-.]+$/u;
  const hasLetter = /\p{L}/u.test(trimmed);

  return allowedPattern.test(trimmed) && hasLetter;
};

export default withErrorBoundary<typeof isValidNameFormat>(isValidNameFormat);

File: packages/i18nify-go/modules/names/is_valid_name_format.go
package names

import (
	"strings"
	"unicode"
)

const (
	minNameLength = 2
	maxNameLength = 100
)

// IsValidNameFormat reports whether name has a structurally valid personal name
// format. It mirrors the JS isValidNameFormat function exactly.
//
// A name has a valid format when it:
//   - Is at least 2 characters long (after trimming whitespace).
//   - Is at most 100 characters long.
//   - Contains only Unicode letters, combining marks (\p{M}), spaces, hyphens
//     (-), apostrophes ('), or periods (.).
//   - Contains at least one Unicode letter.
//
// This is a charset/structure check. For garbage-input detection (blocklists,
// sequential or repeating characters), use IsValidName.
func IsValidNameFormat(name string) bool {
	trimmed := strings.TrimSpace(name)

	runes := []rune(trimmed)
	if len(runes) < minNameLength || len(runes) > maxNameLength {
		return false
	}

	hasLetter := false
	for _, r := range runes {
		switch {
		case unicode.IsLetter(r):
			hasLetter = true
		case unicode.IsMark(r):
			// Combining marks (e.g., Devanagari vowel signs, Arabic diacritics)
			// are integral parts of correctly spelled names in many scripts.
		case r == ' ', r == '-', r == '\'', r == '.':
			// allowed punctuation/separators
		default:
			return false
		}
	}
	return hasLetter
}
